app_dart/lib/src/service/scheduler.dart

```diff
@@ -0,0 +1,224 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:appengine/appengine.dart';
+import 'package:gcloud/db.dart';
+import 'package:googleapis/bigquery/v2.dart';
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart';
+
+import '../datastore/cocoon_config.dart';
+import '../foundation/utils.dart';
+import '../model/appengine/commit.dart';
+import '../model/appengine/task.dart';
+import '../model/devicelab/manifest.dart';
+import '../request_handling/exceptions.dart';
+import '../service/luci.dart';
+import 'datastore.dart';
+
+/// Scheduler service to validate all commits to supported Flutter repositories.
+///
+/// Scheduler responsibilties include:
+///   1. Tracking commits in Cocoon
+///   2. Ensuring commits are validated (via scheduling tasks against commits)
+///   3. Retry mechanisms for taskss
+class Scheduler {
+  Scheduler({
+    @required this.config,
+    @required this.datastore,
+    this.gitHubBackoffCalculator = twoSecondLinearBackoff,
+    this.httpClient,
+    this.log,
+  })  : assert(datastore != null),
+        assert(gitHubBackoffCalculator != null);
+
+  final Config config;
+  final DatastoreService datastore;
+  final HttpClient httpClient;
+  final GitHubBackoffCalculator gitHubBackoffCalculator;
+  final Logging log;
+
+  /// Ensure [commits] exist in Cocoon.
+  ///
+  /// If [Commit] does not exist in Datastore:
+  ///   * Write it to datastore
+  ///   * Schedule tasks listed in its scheduler config
+  /// Otherwise, ignore it.
+  ///
+  /// [commits] is assumed to be sorted in descending order by merged timestamp.
+  Future<void> addCommits(List<Commit> commits) async {
+    final List<Commit> newCommits = await _getNewCommits(commits);
+    log.debug('Found ${newCommits.length} new commits on GitHub');
+    for (Commit commit in newCommits) {
+      await _addCommit(commit);
+    }
+  }
+
+  Future<void> _addCommit(Commit commit) async {
+    try {
+      await datastore.withTransaction<void>((Transaction transaction) async {
+        transaction.queueMutations(inserts: <Commit>[commit]);
+        await transaction.commit();
+        log.debug('Committed commit ${commit.sha}');
+      });
+      await _scheduleTasks(commit);
+    } catch (error) {
+      log.error('Failed to add commit ${commit.sha}: $error');
+    }
+
+    await _uploadToBigQuery(commit);
+  }
+
+  /// Return subset of [commits] not stored in Datastore.
+  Future<List<Commit>> _getNewCommits(List<Commit> commits) async {
+    final List<Commit> newCommits = <Commit>[];
+    // Ensure commits are sorted from newest to oldest
+    commits.sort((Commit a, Commit b) => a.timestamp.compareTo(b.timestamp));
```

Comment:
list is already sorted.

Reply:
This is mostly for tests. I removed documentation. In the next PR, I want to switch this from a list to a single commit and will be irrelevant.